tools/encryption/encryptionutils/encryption-helper.go

```diff
@@ -0,0 +1,243 @@
+		var content []string
+		for c := node.FirstChild; c != nil; c = c.NextSibling {
+			content = append(content, renderNode(c, true))
+			node.RemoveChild(c)
+		}
+		encrypted_content, encrypt_err := cipher.Encrypt([]byte(strings.Join(content, "")), nil)
+		if encrypt_err != nil {
+			return encrypt_err
+		}
+		text_node := &html.Node{Type: html.TextNode, Data: base64.StdEncoding.EncodeToString(encrypted_content)}
+	    attrs := []html.Attribute{
+			html.Attribute{Key: "type", Val: "application/octet-stream"},
+			html.Attribute{Key: "ciphertext", Val: ""},
+		}
+		script_node := &html.Node{
+			Type: html.ElementNode,
+			Data:"script",
+			DataAtom: atom.Script, 
+			Attr: attrs,
+		}
+		node.AppendChild(script_node)
+		script_node.AppendChild(text_node)
+	}
+	return nil
+}
+
+// Retrieves Google's public key from the given URL.
+func getGooglePublicKey(public_key_url string) (tinkpb.Keyset, error) {
+	resp, err := http.Get(public_key_url)
+	if err != nil {
+		return tinkpb.Keyset{}, err
+	}
+	r := keyset.NewJSONReader(resp.Body)
+	ks, err := r.Read()
+	if err != nil {
+		return tinkpb.Keyset{}, err
+	}
+	return *ks, nil
+}
+
+// Encrypts the document's symmetric key using the input Keyset.
+func encryptDocumentKey(doc_keyset string, access_requirement string, ks tinkpb.Keyset) (string, error) {
+	handle, err := keyset.NewHandleWithNoSecrets(&ks)
+	if err != nil {
+		return "", err
+	}
+	he, err := hybrid.NewHybridEncrypt(handle)
+    if err != nil {
+        return "", err
+	}
+	json_str := fmt.Sprintf("{\"accessRequirements\": [\"%s\"], \"key\": \"%s\"}", access_requirement, doc_keyset)
+	enc, err := he.Encrypt([]byte(json_str), nil)
+	if err != nil {
+		return "", err
+	}
+	return base64.StdEncoding.EncodeToString(enc), nil
+}
+
+// Adds the encrypted document key to the output document's head.
+func addEncryptedDocumentKeyToHead(encrypted_key string, parsed_html *html.Node) error {
+	for n := parsed_html.FirstChild; n != nil; n = n.NextSibling {
+		if (n.Data == "html") && (len(n.Attr) != 0) {
+			for cn := n.FirstChild; cn != nil; cn = cn.NextSibling {
+				if (cn.Data == "head") {
+					attrs := []html.Attribute{
+						html.Attribute{Key: "type", Val: "application/json"},
+						html.Attribute{Key: "cryptokeys", Val: ""},
+					}
+					crypto_keys := &html.Node{
+						Type: html.ElementNode,
+						Data:"script",
+						DataAtom: atom.Script, 
+						Attr: attrs,
+					}
+					jsonData := fmt.Sprintf(`{"google.com":"%s"}`, encrypted_key)
+					text_node := &html.Node{Type: html.TextNode, Data: jsonData}
+					crypto_keys.AppendChild(text_node)
+					cn.AppendChild(crypto_keys)
+			    	return nil
+				}
+			}
+		}
+	}
+	return fmt.Errorf("Could not add cryptokeys to head.")
+}
+
+// Renders the input Node to a string.
+func renderNode(n *html.Node, trim bool) string {
+	var buf bytes.Buffer
+	w := io.Writer(&buf)
+	html.Render(w, n)
+	s := buf.String()
+	if trim {
+		s = strings.TrimPrefix(s, "<html><body>")
```

Comment:
Don't think this will catch everything. What if html element has attributes?

Reply:
I have to remove these strings from the output because html.Render adds them in automatically.

Do you know of a way to render the content inside of the html.Node that doesn't add any extra tags?

Reply:
Parse() is what adds the missing tags. Render on a node should simply render that node (and any of its children).

If you run this script, it outputs "<p>foobar</p>".

https://play.golang.org/p/RkjIaSEMGtK

Reply:
Thanks! Switched to Render and removed the trim option.